### PR TITLE
Added state to oauth_url function

### PR DESCRIPTION
**Pull request by Dhruvacube**
## Summary

Added state to oauth_url function, sometimes state is require in oauth url like this https://discord.com/oauth2/authorize?client_id=935242576343224352&permissions=8&scope=applications.commands%20bot&response_type=code&state=cube12345%3F%2FDirect%20From%20Bot&redirect_uri=https%3A%2F%2Fminatonamikaze-invites.herokuapp.com%2Finvite

## Checklist

- [ ] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/utils.py [hunk 1]
```diff
@@ -288,13 +288,14 @@ def oauth_url(
     redirect_uri: str = MISSING,
     scopes: Iterable[str] = MISSING,
     disable_guild_select: bool = False,
+    state: str = MISSING
```
Comment (Rapptz) [inline discord/utils.py:291]:
```suggestion
    state: str = MISSING,
```

File: discord/utils.py [hunk 2]
```diff
@@ -288,13 +288,14 @@ def oauth_url(
...
 ) -> str:
     """A helper function that returns the OAuth2 URL for inviting the bot
     into guilds.
 
     .. versionchanged:: 2.0
 
-        ``permissions``, ``guild``, ``redirect_uri``, and ``scopes`` parameters
+        ``permissions``, ``guild``, ``redirect_uri``, ``scopes`` and ``state`` parameters
```
Comment (Rapptz) [inline discord/utils.py:298]:
```suggestion
        ``permissions``, ``guild``, ``redirect_uri``, and ``scopes`` parameters
```

This isn't a v1.7 -> v2 change.

File: discord/utils.py [hunk 3]
```diff
@@ -334,6 +339,8 @@ def oauth_url(
         url += '&response_type=code&' + urlencode({'redirect_uri': redirect_uri})
     if disable_guild_select:
         url += '&disable_guild_select=true'
+    if state is not MISSING:
+        url += f'&state={state}'
```
Comment (Rapptz) [inline discord/utils.py:343]:
```suggestion
        encoded = urlencode({'state': state})
        url += f'&{encoded}'
```

**Review by Rapptz — CHANGES_REQUESTED** [review]
Thanks for the PR.

I'm not really sure this is going to get much use in the library considering the library doesn't support OAuth2 routes but if this were to get added these changes would need to be made.

File: discord/utils.py (hunk 3 above)
Comment (Dhruvacube) [reply discord/utils.py:343]: Ahh, I totally forgot about the encoding part

File: discord/utils.py (hunk 2 above)
Comment (Dhruvacube) [reply discord/utils.py:298]: ahhh, ohh okay

**Comment by Dhruvacube** [comment]
> Thanks for the PR.
> 
> I'm not really sure this is going to get much use in the library considering the library doesn't support OAuth2 routes but if this were to get added these changes would need to be made.

yep, but as i was rewriting the bot , i thought that i use this use function but the state parameter was missing, so i thought that i might complete it to fulfill my requirements.

File: discord/utils.py [hunk 4]
```diff
@@ -334,6 +339,9 @@ def oauth_url(
         url += '&response_type=code&' + urlencode({'redirect_uri': redirect_uri})
     if disable_guild_select:
         url += '&disable_guild_select=true'
+    if state is not MISSING:
+        encoded = urlencode({'state': state})
```
Comment (Rapptz) [inline discord/utils.py:343]: Looks like the import would have to be moved. I didn't realise it was a function local import.

Comment (Dhruvacube) [reply discord/utils.py:343]: yeah lol wait, i will fix that asap as i just simple commiting the sugesstion

**Review by Rapptz — CHANGES_REQUESTED** [review]
This PR is failing `black` now.

File: discord/utils.py [hunk 5]
```diff
@@ -328,12 +333,15 @@ def oauth_url(
...
         url += f'&guild_id={guild.id}'
-    if redirect_uri is not MISSING:
-        from urllib.parse import urlencode
-
-        url += '&response_type=code&' + urlencode({'redirect_uri': redirect_uri})
     if disable_guild_select:
         url += '&disable_guild_select=true'
+    if redirect_uri is not MISSING or state is not MISSING:
+        from urllib.parse import urlencode
+
```
Comment (GoogleGenius) [inline discord/utils.py:340]:
```suggestion

```

Comment (GoogleGenius) [reply discord/utils.py:340]: Why am I manually looking for whitespace… anyway.

Comment (Dhruvacube) [reply discord/utils.py:340]: lol

**Comment by Dhruvacube** [comment]
> This PR is failing `black` now.

hmmmmmmm idk why maybe debugging time

**Comment by Rapptz** [comment]
It's not a bug in the code, it's a formatting error. This repository uses CI (Continuous Integration) to ensure that all library files are formatted properly. It uses a program called [black](https://github.com/psf/black) to accomplish this. You can download it by doing `pip install -U black`.

The settings for black are located in the root directory of the project where the `pyproject.toml` is located. Make sure to run `black discord/ examples/` in the directory where the `pyproject.toml` is located and it should all fix the formatting issues.

**Comment by Dhruvacube** [comment]
> It's not a bug in the code, it's a formatting error. This repository uses CI (Continuous Integration) to ensure that all library files are formatted properly. It uses a program called [black](https://github.com/psf/black) to accomplish this. You can download it by doing `pip install -U black`.
> 
> The settings for black are located in the root directory of the project where the `pyproject.toml` is located. Make sure to run `black discord/ examples/` in the directory where the `pyproject.toml` is located and it should all fix the formatting issues.

ohhh ok

**Comment by Dhruvacube** [comment]
@Rapptz fixed the formatting issues :) maybe check now?

File: discord/utils.py [hunk 6]
```diff
@@ -626,7 +634,7 @@ async def maybe_coroutine(f: MaybeAwaitableFunc[P, T], *args: P.args, **kwargs:
 
 async def async_all(gen: Iterable[Awaitable[T]], *, check: Callable[[T], bool] = _isawaitable) -> bool:
     for elem in gen:
-        if check(elem):
+        if check(elem): # type: ignore
```
Comment (Rapptz) [inline discord/utils.py:637]: Please remove this.

**Review by Rapptz — CHANGES_REQUESTED** [review]
Sorry, this is still failing `black`.

File: discord/utils.py (hunk 6 above)
Comment (Dhruvacube) [reply discord/utils.py:637]: ok

**Comment by Dhruvacube** [comment]
> Sorry, this is still failing `black`.

ya just saw now

File: discord/utils.py (hunk 6 above)
Comment (Dhruvacube) [reply discord/utils.py:637]: but was giving me a type error

Comment (Dhruvacube) [reply discord/utils.py:637]: image
Here this is the error that vscode was giving

Comment (Rapptz) [reply discord/utils.py]: It was already fixed in a separate PR. Please keep PRs to a single issue.

Comment (Dhruvacube) [reply discord/utils.py]: ohhh ok !!

**Comment by Dhruvacube** [comment]
finally all checks passed !

**Review by Rapptz — APPROVED** [review]
👍

**Pull request merged**